File: 0042-trapping-rain-water/0042-trapping-rain-water.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        int left = 0;
        int right = height.size() - 1;

        int maxL = height[left];
        int maxR = height[right];

        int water = 0;

        while(left < right) {
            if(maxL < maxR) {
                water += maxL - height[left++];
                maxL = max(maxL, height[left]);
            } else {
                water += maxR - height[right--];
                maxR = max(maxR, height[right]);
            }
        }

        return water;
    }
};

File: 0042-trapping-rain-water/0042-trapping-rain-water.ts
function trap(height: number[]): number {
    
    let left     = 0;
    let leftMax  = height[left];
    let right    = height.length - 1;
    let rightMax = height[right];

    let water = 0;

    while(left < right) {
        if(leftMax < rightMax) {
            leftMax = Math.max(leftMax, height[++left]);
            water += leftMax - height[left];
        } else {
            rightMax = Math.max(rightMax, height[--right]);
            water += rightMax - height[right];
        }
    }

    return water;
};